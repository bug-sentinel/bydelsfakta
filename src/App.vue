<template>
  <div id="app" class="app" tabindex="-1" :class="{ menuIsOpen, navigationIsOpen }">
    <pkt-icons-sprite />
    <the-navigation-drawer />
    <div class="app__content">
      <div v-if="showWarning" class="ok-alert-container">
        <div class="ok-alert ok-alert--warning">
          {{ $t('alert.warning') }}
        </div>
      </div>
      <the-navigation-topbar />
      <main>
        <router-view />
      </main>
      <the-footer />
    </div>
    <modal v-if="isOpen" @close="isOpen = false" />
  </div>
</template>

<script>
import { mapState, mapActions } from 'vuex';
import PktIconsSprite from './components/PktIconsSprite.vue';
import TheNavigationDrawer from './components/TheNavigationDrawer.vue';
import TheNavigationTopbar from './components/TheNavigationTopbar.vue';
import TheFooter from './components/TheFooter.vue';
import Modal from './components/Modal.vue';

export default {
  name: 'App',

  components: {
    PktIconsSprite,
    TheNavigationTopbar,
    TheNavigationDrawer,
    TheFooter,
    Modal,
  },

  data: () => ({
    isOpen: false,
    showWarning: false,
    warningMessage: '',
  }),

  metaInfo() {
    return {
      title: 'Bydelsfakta – Visualisering av statistikk om befolkning, levekår og boforhold',
    };
  },

  computed: {
    ...mapState(['menuIsOpen', 'navigationIsOpen', 'ie11']),
  },

  watch: {
    $route(to) {
      if (to.params.district !== undefined) {
        this.addDistrict({ district: to.params.district, pushRoute: false });
        document.getElementById('app').focus();
      }
    },
  },

  mounted() {
    if ('ontouchstart' in document.documentElement) {
      this.setTouchDevice(true);
    }
    if (!!window.MSInputMethodContext && !!document.documentMode) {
      this.isOpen = true;
      this.setIE11Compatibility(true);
    }
  },

  created() {
    if (import.meta.env.PROD) {
      const envs = JSON.parse(window.__GLOBAL_ENVS__);
      this.setProductionMode(envs.VITE_PRODUCTION_DATA);
      document.getElementById('bydelsfakta-globals').remove();
    } else {
      this.setProductionMode(import.meta.env.VITE_PRODUCTION_DATA);
    }

    if (this.$route.name === 'Home') {
      this.addDistrict({ district: 'alle', pushRoute: false });
    }
    if (this.$route.params.district === undefined) {
      return;
    }
    this.addDistrict({ district: this.$route.params.district, pushRoute: true });
  },

  methods: {
    ...mapActions(['addDistrict', 'setTouchDevice', 'setIE11Compatibility', 'setProductionMode']),
  },
};

// Using a class on body to determine how to style focus states
document.body.addEventListener('mousedown', () => {
  document.body.classList.add('using-mouse');
});
document.body.addEventListener('keydown', () => {
  document.body.classList.remove('using-mouse');
});
</script>

<style lang="scss">
$borderRadius: 1.5px;

@mixin boxShadow($color) {
  box-shadow: 0 1px 3px rgba($color, 0.2), 0 3px 12px rgba($color, 0.05);
}

@mixin boxColor($color, $shadowColor: #292858) {
  background: rgba($color, 0.1);
  @include boxShadow($shadowColor);

  &::after {
    background: linear-gradient($color, darken($color, 5%));
  }

  &::before {
    background: $color;
  }
}

.app:focus {
  outline: none;
}

.ok-alert-container {
  display: flex;
  align-items: center;
  justify-content: center;
}

.ok-alert {
  position: relative;
  max-width: 600px;
  margin: 1rem 1.5rem 1rem;
  padding: 1rem 2.5rem 1rem 1.5rem;
  overflow: hidden;
  font-size: 1rem;
  border-radius: $borderRadius;

  &::after {
    position: absolute;
    top: 0;
    right: none;
    bottom: 0;
    left: 0;
    display: inline-flex;
    align-items: inherit;
    justify-content: center;
    width: 0.25rem;
    height: 100%;
    padding-top: 1rem;
    padding-left: 0;
    font-size: 1.75rem;
    font-family: 'Oslo Icons', sans-serif;
    text-align: center;
    content: '';
  }

  @include boxColor(#242433);

  &--warning {
    @include boxColor(#f8c66b, darken(#f8c66b, 20%));
  }
}
</style>
