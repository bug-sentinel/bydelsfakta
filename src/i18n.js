import Vue from 'vue';
import VueI18n from 'vue-i18n';

function loadLocaleMessages() {
  const locales = require('./locales/nb-no.json');
  return {
    'nb-no': locales,
  };
}

export default function setupI18n() {
  Vue.use(VueI18n);

  return new VueI18n({
    locale: process.env.VUE_APP_I18N_LOCALE || 'nb-no',
    fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'nb-no',
    messages: loadLocaleMessages(),
  });
}
