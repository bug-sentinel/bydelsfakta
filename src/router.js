import { createRouter, createWebHistory } from 'vue-router';

import { topicNames, disabledTopics } from './config/topics';
import allDistricts from './config/allDistricts';

export const routes = [
  {
    path: '/',
    name: 'Home',
    component: () => import('@/views/District.vue'),
  },
  {
    path: '/bydel/:district?',
    name: 'District',
    component: () => import('@/views/District.vue'),
    props: true,
  },
  {
    path: '/bydel/:district/:topic',
    name: 'Topic',
    component: () => import('@/views/Topic.vue'),
    props: true,
  },
  {
    path: '/:pathMatch(.*)*',
    name: 'NotFound',
    component: () => import('@/views/NotFound.vue'),
  },
];

const router = createRouter({
  history: createWebHistory(),
  routes,
});

// Temp while we find out what our homepage should look like
router.beforeEach((to, from) => {
  if (to.params.topic) {
    if (
      (!topicNames.find((name) => name === to.params.topic) && import.meta.env.VITE_PRODUCTION_DATA === 'prod') ||
      disabledTopics.includes(to.params.topic)
    ) {
      return { name: 'NotFound', params: [to.path] };
    }
  }
  if (to.params.district) {
    const districts = to.params.district.split('-');
    if (districts[0] === 'alle') {
      return true;
    }
    if (districts.length === 1) {
      if (allDistricts.find((district) => district.uri === districts[0])) {
        return true;
      }
      if (allDistricts.find((district) => district.key === districts[0])) {
        return true;
      }
      return { name: 'NotFound', params: [to.path] };
    }
    if (districts.length > 1) {
      const errors = [];
      districts.forEach((district) => {
        if (!allDistricts.find((item) => item.key === district)) {
          errors.push(district);
        }
      });
      if (errors.length > 0) {
        return { name: 'NotFound', params: [to.path] };
      }
    }
  }
  if (to.params.topic !== from.params.topic || (to.name === 'Topic' && from.name === 'District')) {
    window.scroll(0, 0);
  }
  return true;
});

export default router;
