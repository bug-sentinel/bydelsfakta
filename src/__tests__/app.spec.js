import { mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import { createRouter, createWebHistory } from 'vue-router';
import Vue3Resize from 'vue3-resize';
import VueSkipTo from '@vue-a11y/skip-to';
import App from '@/App.vue';
import clickOutside from '@/directives/clickOutside';
import i18n from '@/i18n';
import mockStore from '@/../tests/MockStore';
import { routes } from '@/router';

global.scroll = jest.fn();
window.scroll = jest.fn();

describe('App', () => {
  let wrapper = null;
  let store = null;
  let router = null;

  beforeEach(() => {
    store = createStore(mockStore);
    router = createRouter({
      history: createWebHistory(),
      routes,
    });

    wrapper = mount(App, {
      global: {
        plugins: [router, store, i18n, Vue3Resize, VueSkipTo],
        directives: {
          'click-outside': clickOutside,
        },
      },
      shallow: true,
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders app-component and finds id #app', () => {
    expect(wrapper.classes('app')).toBe(true);
  });

  test('shallowmounts app', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
