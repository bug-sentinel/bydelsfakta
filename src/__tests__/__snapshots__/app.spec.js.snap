// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App shallowmounts app 1`] = `
<div
  class="app menuIsOpen"
  id="app"
  tabindex="-1"
>
  <pkt-icons-sprite-stub />
  <the-navigation-drawer-stub />
  <div
    class="app__content"
  >
    <!--v-if-->
    <the-navigation-topbar-stub />
    <main>
      <router-view-stub
        name="default"
      />
    </main>
    <the-footer-stub />
  </div>
  <!--v-if-->
</div>
`;
